Sources/HTTP/HTTPServer.swift

```diff
@@ -1,54 +1,336 @@
+    var queues = [ DispatchQueue ]()
+    for i in 0..<ProcessInfo.processInfo.processorCount {
+      let q = DispatchQueue(label: "de.zeezide.swift.server.http.handler\(i)")
+      queues.append(q)
+      q.async { _ = 46 + 2 } // warm up threads
+    }
+    return queues
+  }()
+
+  /// Queue used to run `accept()` on.
+  /// `accept()` can take time (AFAIK there is no NIO variant of it), run it
+  /// multithreaded. GCD will decide the thread count and such for us.
+  private let acceptQueue =
+                DispatchQueue(label: "de.zeezide.swift.server.http.accept",
+                              attributes: .concurrent)
+  
+  public init() {
+  }
+  deinit {
+    stop()
+  }
+  
+  
+  /// Start the HTTP server on the given `port` number, using a
+  /// `HTTPRequestHandler` to process incoming requests.
+  public func start(port: Int? = nil,
+                    handler: @escaping HTTPRequestHandler) throws
+  {
+    // - port as Int=0 vs Int? - Int? is better design
+    try start(address: sockaddr_in(port: port), handler: handler)
+  }
+  
+  /// Start the HTTP server on the given `port` number, using a
+  /// `HTTPRequestHandler` to process incoming requests.
+  func start(address : sockaddr_in,
+             backlog : Int = 511,
+             handler : @escaping HTTPRequestHandler) throws
+  {
+    /* setup socket */
+    
+    let ( fd, address ) = try createSocket(boundTo: address)
+    boundAddress = address
+    
+    /* Setup Listen Source */
+
+    listenSource = DispatchSource.makeReadSource(fileDescriptor: fd,
+                                                 queue: acceptQueue)
+    listenSource?.setEventHandler {
+      // if we don't pass it over, we need to synchronize access to the fd!
+      self.handleListenEvent(on: fd, handler: handler, localAddress: address)
+    }
+    
+    listenSource?.resume()
+    
+    /* Listen */
+    
+    let rc = listen(fd, Int32(backlog))
+    if rc != 0 {
+      let error = errno
+      listenSource?.cancel()
+      listenSource = nil
+      close(fd)
+      throw Error.socketError(.listenFailed(error))
+    }
+  }
+  
+  public func stop() {
+    // TBD: argument: `hard: Bool`, similar to DispatchIO
+    // TBD: hard stop vs Apache-like 'let requests finish'
+    
+    if let source = listenSource {
+      source.cancel()
+      close(Int32(source.handle))
+      source.setEventHandler(handler: nil)
+      listenSource = nil
+    }
+    boundAddress = nil
+    
+    queue.async {
+      self.connections.forEach { $0.serverWillStop() }
+    }
+  }
+  
+  public var port : Int {
+    guard let boundAddress = boundAddress else { return -1 }
+    return Int(boundAddress.sin_port.littleEndian)
+  }
+  
+  
+  private(set) public var connectionCount : Int32 = 0
+  
+  
+  private func handleListenEvent(on fd        : Int32,
+                                 handler      : @escaping HTTPRequestHandler,
+                                 localAddress : sockaddr_in)
+  {
```

Comment:
In general I prefer:
```.swift
private func handleListenEvent(
  on fd: Int32,
  handler      : @escaping HTTPRequestHandler,
  localAddress : sockaddr_in
) {
```

Reply:
I don't, but I don't care much either.

Reply:
Not that I care a lot but I only know this style guide: https://github.com/raywenderlich/swift-style-guide and that doesn't have it ...